jetstream.go

```diff
@@ -0,0 +1,334 @@
+package nats
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	AckAck      = []byte("+ACK")
+	AckNak      = []byte("-NAK")
+	AckProgress = []byte("+WPI")
+	AckNext     = []byte("+NXT")
+	AckTerm     = []byte("+TERM")
+)
+
+// JetStreamMsgMetaData is metadata related to a JetStream originated message
+type JetStreamMsgMetaData struct {
+	stream    string
+	consumer  string
+	parsed    bool
+	delivered int
+	sSeq      int
+	cSeq      int
+	ts        time.Time
+}
+
+func (m *JetStreamMsgMetaData) Stream() string {
+	return m.stream
+}
+
+func (m *JetStreamMsgMetaData) Consumer() string {
+	return m.consumer
+}
+
+func (m *JetStreamMsgMetaData) Valid() bool {
+	return m.parsed
+}
+
+func (m *JetStreamMsgMetaData) Delivered() int {
+	return m.delivered
+}
+
+func (m *JetStreamMsgMetaData) StreamSequence() int {
+	return m.sSeq
+}
+
+func (m *JetStreamMsgMetaData) ConsumerSequence() int {
+	return m.cSeq
+}
+
+func (m *JetStreamMsgMetaData) PublishTime() time.Time {
+	return m.ts
+}
+
+func (m *Msg) JetStreamMetaData() (*JetStreamMsgMetaData, error) {
+	var err error
+
+	if m.jsMeta != nil && m.jsMeta.parsed {
+		return m.jsMeta, nil
+	}
+
+	m.jsMeta, err = m.parseJSMsgMetadataUnlocked()
+
+	return m.jsMeta, err
+}
+
+func (m *Msg) parseJSMsgMetadataUnlocked() (*JetStreamMsgMetaData, error) {
+	if m.jsMeta != nil {
+		return m.jsMeta, nil
+	}
+
+	if len(m.Reply) == 0 {
+		return nil, ErrNotJSMessage
+	}
+
+	meta := &JetStreamMsgMetaData{}
+
+	parts := strings.Split(m.Reply, ".")
```

Comment:
Could handroll this one as well for performance. We have this for the proto parser and subjects I believe.

Reply:
hmm, cant find that example, found some in `processMsgArgs` but basing on those performed worse than string.Split, mind pointing at the one you had in mind?

Reply:
Can't find it for reference, but here it is..

```go
	// Tokenize the subject. This should always be a literal subject.
	tsa := [32]string{}
	tts := tsa[:0]
	start := 0
	for i := 0; i < len(subject); i++ {
		if subject[i] == btsep {
			tts = append(tts, subject[start:i])
			start = i + 1
		}
	}
	tts = append(tts, subject[start:])
```